x/distribution/genesis.go

```diff
@@ -11,30 +11,87 @@ func InitGenesis(ctx sdk.Context, keeper Keeper, data types.GenesisState) {
 	keeper.SetCommunityTax(ctx, data.CommunityTax)
 	keeper.SetBaseProposerReward(ctx, data.BaseProposerReward)
 	keeper.SetBonusProposerReward(ctx, data.BonusProposerReward)
-
-	for _, vdi := range data.ValidatorDistInfos {
-		keeper.SetValidatorDistInfo(ctx, vdi)
+	for _, dwi := range data.DelegatorWithdrawInfos {
+		keeper.SetDelegatorWithdrawAddr(ctx, dwi.DelegatorAddr, dwi.WithdrawAddr)
 	}
-	for _, ddi := range data.DelegationDistInfos {
-		keeper.SetDelegationDistInfo(ctx, ddi)
+	keeper.SetPreviousProposerConsAddr(ctx, data.PreviousProposer)
+	keeper.SetOutstandingRewards(ctx, data.OutstandingRewards)
+	for _, acc := range data.ValidatorAccumulatedCommissions {
+		keeper.SetValidatorAccumulatedCommission(ctx, acc.ValidatorAddr, acc.Accumulated)
 	}
-	for _, dw := range data.DelegatorWithdrawInfos {
-		keeper.SetDelegatorWithdrawAddr(ctx, dw.DelegatorAddr, dw.WithdrawAddr)
+	for _, his := range data.ValidatorHistoricalRewards {
+		keeper.SetValidatorHistoricalRewards(ctx, his.ValidatorAddr, his.Period, his.Rewards)
+	}
+	for _, cur := range data.ValidatorCurrentRewards {
+		keeper.SetValidatorCurrentRewards(ctx, cur.ValidatorAddr, cur.Rewards)
+	}
+	for _, del := range data.DelegatorStartingInfos {
+		keeper.SetDelegatorStartingInfo(ctx, del.ValidatorAddr, del.DelegatorAddr, del.StartingInfo)
+	}
+	for _, evt := range data.ValidatorSlashEvents {
+		keeper.SetValidatorSlashEvent(ctx, evt.ValidatorAddr, evt.Height, evt.Event)
 	}
-	keeper.SetPreviousProposerConsAddr(ctx, data.PreviousProposer)
 }
 
-// ExportGenesis returns a GenesisState for a given context and keeper. The
-// GenesisState will contain the pool, and validator/delegator distribution info's
+// ExportGenesis returns a GenesisState for a given context and keeper.
 func ExportGenesis(ctx sdk.Context, keeper Keeper) types.GenesisState {
 	feePool := keeper.GetFeePool(ctx)
 	communityTax := keeper.GetCommunityTax(ctx)
 	baseProposerRewards := keeper.GetBaseProposerReward(ctx)
 	bonusProposerRewards := keeper.GetBonusProposerReward(ctx)
-	vdis := keeper.GetAllValidatorDistInfos(ctx)
-	ddis := keeper.GetAllDelegationDistInfos(ctx)
-	dwis := keeper.GetAllDelegatorWithdrawInfos(ctx)
+	dwi := make([]types.DelegatorWithdrawInfo, 0)
+	keeper.IterateDelegatorWithdrawAddrs(ctx, func(del sdk.AccAddress, addr sdk.AccAddress) (stop bool) {
+		dwi = append(dwi, types.DelegatorWithdrawInfo{
+			DelegatorAddr: del,
+			WithdrawAddr:  addr,
+		})
+		return false
+	})
 	pp := keeper.GetPreviousProposerConsAddr(ctx)
-	return NewGenesisState(feePool, communityTax, baseProposerRewards,
-		bonusProposerRewards, vdis, ddis, dwis, pp)
+	outstanding := keeper.GetOutstandingRewards(ctx)
```

Comment:
Perhaps we should move this call to the bottom where we return (when we actually need it).